app/src/cose.h

```diff
@@ -207,6 +439,388 @@ namespace scitt::cose
+  {
+    int key_len_bits;
+    unsigned key_len_bytes;
+
+    key_len_bits = EVP_PKEY_bits(key_evp);
+
+    /* Calculation of size per RFC 8152 section 8.1 -- round up to
+     * number of bytes. */
+    key_len_bytes = (unsigned)key_len_bits / 8;
+    if (key_len_bits % 8)
+    {
+      key_len_bytes++;
+    }
+
+    return key_len_bytes;
+  }
+
+  // Temporarily needed for notary_verify().
+  std::vector<uint8_t> create_tbs(const std::vector<uint8_t>& cose_sign1)
+  {
+    // Note: This function does not return the hash of the TBS because
+    // EdDSA does not support pre-hashed messages as input.
+
+    // Sig_structure = [
+    //     context: "Signature1",
+    //     body_protected: empty_or_serialized_map,
+    //     external_aad: bstr,
+    //     payload: bstr
+    // ]
+
+    // Extract fields from the COSE_Sign1 message.
+    UsefulBufC msg{cose_sign1.data(), cose_sign1.size()};
+
+    QCBORDecodeContext decode_ctx;
+    QCBORDecode_Init(&decode_ctx, msg, QCBOR_DECODE_MODE_NORMAL);
+
+    QCBORDecode_EnterArray(&decode_ctx, nullptr);
+
+    QCBORItem item;
+
+    // body_protected
+    QCBORDecode_GetNext(&decode_ctx, &item);
+    auto body_protected = item.val.string;
+
+    // skip unprotected header
+    QCBORDecode_VGetNextConsume(&decode_ctx, &item);
+
+    // payload
+    QCBORDecode_GetNext(&decode_ctx, &item);
+    auto payload = item.val.string;
+
+    // signature
+    QCBORDecode_GetNext(&decode_ctx, &item);
+    auto signature = item.val.string;
+
+    QCBORDecode_ExitArray(&decode_ctx);
+    auto error = QCBORDecode_Finish(&decode_ctx);
+    if (error)
+    {
+      throw std::runtime_error("Failed to decode COSE_Sign1");
+    }
+
+    // Create Sig_structure.
+    // Note that hashing the structure incrementally to avoid
+    // doubling memory would work for RSA-PSS and ECDSA, but not EdDSA.
+
+    std::vector<uint8_t> sig_structure_buf(cose_sign1.size() + 1024);
+    UsefulBuf sig_structure{sig_structure_buf.data(), sig_structure_buf.size()};
+
+    QCBOREncodeContext encode_ctx;
+    QCBOREncode_Init(&encode_ctx, sig_structure);
+
+    QCBOREncode_OpenArray(&encode_ctx);
+
+    // context
+    QCBOREncode_AddSZString(&encode_ctx, "Signature1");
+
+    // body_protected: The protected header of the message.
+    QCBOREncode_AddBytes(&encode_ctx, body_protected);
+
+    // external_aad: always empty.
+    QCBOREncode_AddBytes(&encode_ctx, NULLUsefulBufC);
+
+    // payload: The payload of the message.
+    QCBOREncode_AddBytes(&encode_ctx, payload);
+
+    QCBOREncode_CloseArray(&encode_ctx);
+
+    UsefulBufC encoded_cbor;
+    QCBORError err;
+    err = QCBOREncode_Finish(&encode_ctx, &encoded_cbor);
+    if (err != QCBOR_SUCCESS)
+    {
+      throw std::runtime_error("Error encoding CBOR");
+    }
+
+    auto cbor = std::vector<uint8_t>(
+      (uint8_t*)encoded_cbor.ptr,
+      (uint8_t*)encoded_cbor.ptr + encoded_cbor.len);
+    return cbor;
+  }
+
+  static enum t_cose_err_t ecdsa_signature_cose_to_der(
```

Comment:
I think we're better off copying from CCF than t_cose for this: https://github.com/microsoft/CCF/blob/5972983bce5e345666e5bf87d0f28ac1f7f05a84/src/crypto/ecdsa.cpp#L14-L45

Annoyingly, this API is public in CCF 4.x but not 3.x.